### PR TITLE
Fellow members fetching

### LINKED ISSUE #61: Obtain handles from the fellows
Related issue: https://github.com/polkadot-fellows/runtimes/issues/7

Blocker: https://github.com/paritytech/polkadot-sdk/issues/179

### PR DESCRIPTION
**Pull request by Bullrich**
Created the method to fetch the fellows from the parachain using Polkadot-JS. This will be updated to the new library once it becomes available.

The system fetches all the available fellows with a GitHub handle and caches them. When requesting a fellow, it automatically filters the one it needs and returns an array of handles.

This resolves #61

Co-authored-by: Yuri Volkov <mutantcornholio@users.noreply.github.com>
Co-authored-by: Maksym Hlukhovtsov <mordamax@users.noreply.github.com>

### CODE REVIEW

File: src/polkadot/fellows.ts
Comment (mordamax) [inline src/polkadot/fellows.ts:90]: this will use data from cache in case if there are more than 1 rule bucket with rank: x?
i don't see any other use case then... right?

File: README.md
Comment (mordamax) [inline README.md:246]: minFellowRank: 2 ?
why is this in allowedToSkipRule? means that if I'm rank: 3 then no reviews needed?

File: src/test/fellows.test.ts
Comment (mordamax) [inline src/test/fellows.test.ts:28]: this is nice test. do you plan to add the actual rule with rank:2 where you see it allows / or disallows if approver is rank < 2 ?

File: src/polkadot/fellows.ts
Comment (Bullrich) [reply src/polkadot/fellows.ts:90]: It is a map. The idea for this is to fetch **all** the available fellows once and then we filter per request, instead of fetching them multiple times per required rank.

So if you request in one rule a rank 4, it will already fetch them all, so, if you request later rank 2, all the ranks will be available already.

File: README.md
Comment (Bullrich) [reply README.md:246]: Oh, I put it in the wrong place by accident. I fixed it now.

File: src/test/fellows.test.ts
Comment (Bullrich) [reply src/test/fellows.test.ts:28]: That's a good question. I didn't plan on adding it, but maybe it's no trouble at all.

Comment (Bullrich) [reply src/test/fellows.test.ts:28]: It actually runs out of the box, so that's good!

File: src/polkadot/fellows.ts
Comment (rzadp) [inline src/polkadot/fellows.ts:80]:
```suggestion
      this.logger.error(error);
```

File: src/polkadot/fellows.ts
Comment (rzadp) [inline src/polkadot/fellows.ts:105]:
```suggestion
    this.logger.info(`GitHub members of rank '${requiredRank}' or higher are: ${users.join(",")}`);
```

This could be a little easier to read maybe.

**Review by rzadp — APPROVED** [review]
Looks good, just one thing: do you have plans for invalidating the cache of fellows?

File: src/polkadot/fellows.ts
Comment (mutantcornholio) [inline src/polkadot/fellows.ts:28]: Wait, is it basically this?
```ts
fellows.push({address: key.toHuman()[0], rank: rank.toHuman()[0]});
```

File: src/polkadot/fellows.ts
Comment (mutantcornholio) [inline src/polkadot/fellows.ts:43]: I feel like this deserves a semaphore, so it could be queried in parallel.

**Review by mutantcornholio — CHANGES_REQUESTED** [review]

File: src/polkadot/fellows.ts
Comment (Bullrich) [reply src/polkadot/fellows.ts]: Good catch!

**Comment by Bullrich** [comment]
> Looks good, just one thing: do you have plans for invalidating the cache of fellows?

No, as the cache is only generated during the instance and not stored long term. The same as with the GitHub reviews, the cache is only used if we need to fetch the same information during the same execution (that lasts less than a minute).

So if rule 1 request the teams, we fetch them and cache them, and if rule 2 requests it, we already have them. But once the execution is over the cache disappears with it.

File: src/polkadot/fellows.ts
Comment (Bullrich) [reply src/polkadot/fellows.ts:43]: Do you know if there is a limit to stop abuse? That was my main concern.

The current execution takes about 2s, which given the fact that this is an action, is not really a problem (the whole action is executed in about 5s).

File: src/polkadot/fellows.ts
Comment (Bullrich) [reply src/polkadot/fellows.ts]: As you fixed it in the previous comment, I originally had the api.disconnect in there.

File: src/polkadot/fellows.ts
Comment (mutantcornholio) [reply src/polkadot/fellows.ts:43]: > Do you know if there is a limit to stop abuse? That was my main concern.

Could you elaborate?

Comment (Bullrich) [reply src/polkadot/fellows.ts:43]: Same problem that we have with GitHub, which has a rate limit.

I assume that there is some kind of rate limit when using Polkadot-JS (else what stop people from DOSing it?), and, because I'm fetching 60 accounts, I found it better to do it in order.

Comment (Bullrich) [reply src/polkadot/fellows.ts:43]: Maybe I'm not understanding you. Could you please elaborate on what you had in mind?

Comment (mutantcornholio) [reply src/polkadot/fellows.ts:43]: I think we're speaking of the same thing here. I suggest adding a semaphore, so requests would be processed in parallel, but with a limit on how much parallel requests to make. 

In `gitspiegel`, I've used [semaphore-promise](https://github.com/Granjow/semaphore-promise) library for that.  
Check out the usage in [core.ts](https://github.com/paritytech/gitspiegel/blob/master/src/core.ts#L22C7-L22C25)

File: src/test/fellows.test.ts
Comment (mordamax) [reply src/test/fellows.test.ts:28]: > It actually runs out of the box, so that's good!

where? I haven't seen in this diff changes where the rank: 2 approvals are being tested, could you please point my eyes?

Comment (Bullrich) [reply src/test/fellows.test.ts:28]: I added a test to [`src/test/runner/runner.test.ts`](https://github.com/paritytech/review-bot/pull/79/files#diff-22bfdc6c782b01f44253ec2a1bda5c07281909868cfcad6e4f0a79f81b30651dR58) where it verifies that the user is of a given rank and skips the test.

Comment (mordamax) [reply src/test/fellows.test.ts:28]: hmm.. so if author is fellow - it means he doesn't need review? or I don't get
I thought they need to check whether someone how has approved - is a fellow with dan >= X, so then it counts. otherwise the review-bot blocks merge. So I was expecting this kind of test

**Review by mordamax — APPROVED** [review]

**Pull request merged**
